src/rollsum.c

```diff
@@ -26,26 +26,28 @@
 #define DO4(buf,i)  DO2(buf,i); DO2(buf,i+2);
 #define DO8(buf,i)  DO4(buf,i); DO4(buf,i+4);
 #define DO16(buf)   DO8(buf,0); DO8(buf,8);
-#define OF16(off)  {s1 += 16*off; s2 += 136*off;}
 
 void RollsumUpdate(Rollsum *sum,const unsigned char *buf,unsigned int len) {
-    /* ANSI C says no overflow for unsigned. 
-     zlib's adler 32 goes to extra effort to avoid overflow*/
+    /* ANSI C says no overflow for unsigned. zlib's adler32 goes to extra
+    effort to avoid overflow for its mod prime, which we don't have.*/
+    unsigned int n = len;
     unsigned long s1 = sum->s1;
     unsigned long s2 = sum->s2;
 
-    sum->count+=len;                   /* increment sum count */
     while (len >= 16) {
         DO16(buf);
-        OF16(ROLLSUM_CHAR_OFFSET);
         buf += 16;
         len -= 16;
     }
     while (len != 0) {
-        s1 += (*buf++ + ROLLSUM_CHAR_OFFSET);
+        s1 += *buf++;
         s2 += s1;
         len--;
     }
+    /* Increment s1 and s2 by the amounts added by the char offset. */
+    s1 += n * ROLLSUM_CHAR_OFFSET;
+    s2 += ((n*(n+1))>>1) * ROLLSUM_CHAR_OFFSET;
```

Comment:
If the `>>1` is not specifically because you need some specific shifting behavior for overflow (presumably not given `n` is the length in memory), I think just `/2` would be clearer.